src/jit/codegenxarch.cpp

```diff
@@ -7488,115 +7514,183 @@ void CodeGen::genPutArgStk(GenTreePutArgStk* putArgStk)
+        assert((putArgStk->isPushKind()) && !preAdjustedStack && m_pushStkArg);
+
+        // If we have pre-adjusted the stack and are simply storing the fields in order) set the offset to 0.
+        // (Note that this mode is not currently being used.)
+        // If we are pushing the arguments (i.e. we have not pre-adjusted the stack), then we are pushing them
+        // in reverse order, so we start with the current field offset at the size of the struct arg (which must be
+        // a multiple of the target pointer size).
+        unsigned  currentOffset   = (preAdjustedStack) ? 0 : argSize;
+        unsigned  prevFieldOffset = currentOffset;
+        regNumber tmpReg          = REG_NA;
         for (GenTreeFieldList* current = fieldList; current != nullptr; current = current->Rest())
         {
             GenTree* const fieldNode   = current->Current();
             const unsigned fieldOffset = current->gtFieldOffset;
             var_types      fieldType   = current->gtFieldType;
+            const int      fieldSize   = genTypeSize(fieldType);
 
             // Long-typed nodes should have been handled by the decomposition pass, and lowering should have sorted the
             // field list in descending order by offset.
             assert(!varTypeIsLong(fieldType));
             assert(fieldOffset <= prevFieldOffset);
 
-            // TODO-X86-CQ: If this is not a register candidate, or is not in a register,
-            // make it contained.
-            genConsumeReg(fieldNode);
-
-            // If the field is slot-like, we can store the entire register no matter the type.
-            const bool fieldIsSlot =
-                varTypeIsIntegralOrI(fieldType) && ((fieldOffset % 4) == 0) && ((prevFieldOffset - fieldOffset) >= 4);
+            // If the field is slot-like, we can use a push instruction to store the entire register no matter the type.
+            //
+            // The GC encoder requires that the stack remain 4-byte aligned at all times. Round the adjustment up
+            // to the next multiple of 4. If we are going to generate a `push` instruction, the adjustment must
+            // not require rounding.
+            // NOTE: if the field is of GC type, we must use a push instruction, since the emitter is not otherwise
+            // able to detect stores into the outgoing argument area of the stack on x86.
+            const bool fieldIsSlot = ((fieldOffset % 4) == 0) && ((prevFieldOffset - fieldOffset) >= 4);
+            int        adjustment  = roundUp(currentOffset - fieldOffset, 4);
             if (fieldIsSlot)
             {
-                fieldType = genActualType(fieldType);
-                assert(genTypeSize(fieldType) == 4);
+                fieldType         = genActualType(fieldType);
+                unsigned pushSize = genTypeSize(fieldType);
+                assert((pushSize % 4) == 0);
+                noway_assert((adjustment % TARGET_POINTER_SIZE) == 0);
```

Comment:
This had better be guaranteed by `roundUp`... I would leave this as an `assert` (or remove it entirely).